File: bucket_3/12.ts
import * as grpc from 'grpc';
import { Empty } from 'google-protobuf/google/protobuf/empty_pb';
import {
  CensorTweetsRequest,
  CensorTweetsResponse,
  ExportTweetsRequest,
  ExportTweetsResponse,
  FollowUserRequest,
  LikeTweetRequest,
  RetrieveTweetsRequest,
  RetrieveTweetsResponse,
  SortingEnum,
  SocialGraphServiceService,
} from './social_graph_service_pb';
import {
  CreateTweetRequest,
  DeleteTweetRequest,
  Tweet,
  TweetServiceService,
} from './tweet_service_pb';
import {
  AssociateUserRequest,
  BatchGetUserRequest,
  BatchGetUserResponse,
  CheckUniqueUserNameRequest,
  CheckUniqueUserNameResponse,
  GetUserRequest,
  InitialSignupRequest,
  SearchUserResponse,
  SearchUserRequest,
  UpdateUserRequest,
  User,
  UserServiceService,
} from './user_service_pb';

class SocialGraphServicer implements SocialGraphServiceService.ISocialGraphServiceServer {
  retrieveTweets(request: RetrieveTweetsRequest): Promise<RetrieveTweetsResponse> {
    // Implement logic to retrieve tweets
    const response = new RetrieveTweetsResponse();
    // Replace with appropriate data
    return Promise.resolve(response);
  }

  followUser(request: FollowUserRequest): Promise<Empty> {
    // Implement logic to follow a user
    return Promise.resolve(new Empty());
  }

  likeTweet(request: LikeTweetRequest): Promise<Empty> {
    // Implement logic to like a tweet
    return Promise.resolve(new Empty());
  }

  censorTweets(request: CensorTweetsRequest): Promise<CensorTweetsResponse> {
    // Implement logic to censor tweets
    const response = new CensorTweetsResponse();
    // Replace with appropriate data
    return Promise.resolve(response);
  }

  exportTweets(request: ExportTweetsRequest): Promise<ExportTweetsResponse> {
    // Implement logic to export tweets
    const response = new ExportTweetsResponse();
    // Replace with appropriate data
    return Promise.resolve(response);
  }
}

class TweetServicer implements TweetServiceService.ITweetServiceServer {
  createTweet(request: CreateTweetRequest): Promise<Tweet> {
    // Implement logic to create a tweet
    const response = new Tweet();
    // Replace with appropriate data
    return Promise.resolve(response);
  }

  deleteTweet(request: DeleteTweetRequest): Promise<Tweet> {
    // Implement logic to delete a tweet
    const response = new Tweet();
    // Replace with appropriate data
    return Promise.resolve(response);
  }
}

class UserServicer implements UserServiceService.IUserServiceServer {
  initialSignup(request: InitialSignupRequest): Promise<User> {
    // Implement logic for initial signup
    const response = new User();
    // Replace with appropriate data
    return Promise.resolve(response);
  }

  verifyEmail(request: VerifyEmailRequest): Promise<User> {
    // Implement logic to verify email
    const response = new User();
    // Replace with appropriate data
    return Promise.resolve(response);
  }

  associateUser(request: AssociateUserRequest): Promise<User> {
    // Implement logic to associate users
    const response = new User();
    // Replace with appropriate data
    return Promise.resolve(response);
  }

  updateUser(request: UpdateUserRequest): Promise<User> {
    // Implement logic to update user
    const response = new User();
    // Replace with appropriate data
    return Promise.resolve(response);
  }

  checkUniqueUserName(request: CheckUniqueUserNameRequest): Promise<CheckUniqueUserNameResponse> {
    // Implement logic to check unique username
    const response = new CheckUniqueUserNameResponse();
    // Replace with appropriate data
    return Promise.resolve(response);
  }

  searchUser(request: SearchUserRequest): Promise<SearchUserResponse> {
    // Implement logic to search users
    const response = new SearchUserResponse();
    // Replace with appropriate data
    return Promise.resolve(response);
  }

  getUser(request: GetUserRequest): Promise<User> {
    // Implement logic to get user
    const response = new User();
    // Replace with appropriate data
    return Promise.resolve(response);
  }

  batchGetUser(request: BatchGetUserRequest): Promise<BatchGetUserResponse> {
    // Implement logic to batch get users
    const response = new BatchGetUserResponse();
    // Replace with appropriate data
    return Promise.resolve(response);
  }
}

const socialGraphServer = new grpc.Server();
socialGraphServer.addService(SocialGraphServiceService, new SocialGraphServicer());
socialGraphServer.bind('localhost:50052', grpc.ServerCredentials.createInsecure());

const tweetServer = new grpc.Server();
tweetServer.addService(TweetServiceService, new TweetServicer());
tweetServer.bind('localhost:50053', grpc.ServerCredentials.createInsecure());

const userServer = new grpc.Server();
userServer.addService(UserServiceService, new UserServicer());
userServer.bind('localhost:50054', grpc.ServerCredentials.createInsecure());

socialGraphServer.start();
tweetServer.start();
userServer.start();

File: bucket_3/12.py
import grpc
from google.protobuf import empty_pb2
from social_graph_service_pb2 import *
from social_graph_service_pb2_grpc import SSocialGraphService
from tweet_service_pb2 import *
from tweet_service_pb2_grpc import STweetService
from user_service_pb2 import *
from user_service_pb2_grpc import SUserService

class SocialGraphServicer(SSocialGraphService):
    def RetrieveTweets(self, request, context):
        # Implement logic to retrieve tweets
        response = RetrieveTweetsResponse()
        # Replace with appropriate data
        return response

    def FollowUser(self, request, context):
        # Implement logic to follow a user
        return empty_pb2.Empty()

    def LikeTweet(self, request, context):
        # Implement logic to like a tweet
        return empty_pb2.Empty()

    def CensorTweets(self, request, context):
        # Implement logic to censor tweets
        response = CensorTweetsResponse()
        # Replace with appropriate data
        return response

    def ExportTweets(self, request, context):
        # Implement logic to export tweets
        response = ExportTweetsResponse()
        # Replace with appropriate data
        return response

class TweetServicer(STweetService):
    def CreateTweet(self, request, context):
        # Implement logic to create a tweet
        response = Tweet()
        # Replace with appropriate data
        return response

    def DeleteTweet(self, request, context):
        # Implement logic to delete a tweet
        response = Tweet()
        # Replace with appropriate data
        return response

class UserServicer(SUserService):
    def InitialSignup(self, request, context):
        # Implement logic for initial signup
        response = User()
        # Replace with appropriate data
        return response

    def VerifyEmail(self, request, context):
        # Implement logic to verify email
        response = User()
        # Replace with appropriate data
        return response

    def AssociateUser(self, request, context):
        # Implement logic to associate users
        response = User()
        # Replace with appropriate data
        return response

    def UpdateUser(self, request, context):
        # Implement logic to update user
        response = User()
        # Replace with appropriate data
        return response

    def CheckUniqueUserName(self, request, context):
        # Implement logic to check unique username
        response = CheckUniqueUserNameResponse()
        # Replace with appropriate data
        return response

    def SearchUser(self, request, context):
        # Implement logic to search users
        response = SearchUserResponse()
        # Replace with appropriate data
        return response

    def GetUser(self, request, context):
        # Implement logic to get user
        response = User()
        # Replace with appropriate data
        return response

    def BatchGetUser(self, request, context):
        # Implement logic to batch get users
        response = BatchGetUserResponse()
        # Replace with appropriate data
        return response

def serve():
    social_graph_server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    SSocialGraphService.add_SocialGraphServicer_to_server(SocialGraphServicer(), social_graph_server)
    social_graph_server.add_insecure_port('[::]:50052')
    social_graph_server.start()

    tweet_server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    STweetService.add_TweetServicer_to_server(TweetServicer(), tweet_server)
    tweet_server.add_insecure_port('[::]:50053')
    tweet_server.start()

    user_server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    SUserService.add_UserServicer_to_server(UserServicer(), user_server)
    user_server.add_insecure_port('[::]:50054')
    user_server.start()

    try:
        while True:
            time.sleep(_ONE_DAY_IN_SECONDS)
    except KeyboardInterrupt:
        social_graph_server.stop(0)
        tweet_server.stop(0)
        user_server.stop(0)

if __name__ == '__main__':
    serve()
